code/parachain/frame/composable-maths/src/dex/constant_product.rs

```diff
@@ -185,3 +185,127 @@ pub fn compute_deposit_lp(
 		Ok((overwritten_quote_amount, lp_to_mint))
 	}
 }
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum ConstantProductAmmError {
+	ArithmeticError(ArithmeticError),
+	CannotTakeMoreThanAvailable,
+	InvalidTokensList,
+}
+
+impl From<ArithmeticError> for ConstantProductAmmError {
+	fn from(error: ArithmeticError) -> Self {
+		ConstantProductAmmError::ArithmeticError(error)
+	}
+}
+
+impl From<ConstantProductAmmError> for DispatchError {
+	fn from(error: ConstantProductAmmError) -> Self {
+		match error {
+			ConstantProductAmmError::ArithmeticError(error) => DispatchError::from(error),
+			ConstantProductAmmError::CannotTakeMoreThanAvailable =>
+				DispatchError::from(
+					"`a_out` must not be greater than `b_o` (can't take out more than what's available)!"
+				),
+			ConstantProductAmmError::InvalidTokensList => DispatchError::from("Must provide non-empty tokens list!"),
+		}
+	}
+}
+
+/// Computes the decimal value of a a `PerThing` by taking the deconstructed parts of a `PerThing`
+/// and dividing them by `PerThing::one()`.
+///
+/// # Example
+/// ```rust,ignore
+/// let per_thing = PerMill::from_percent(10);
+/// assert_eq!(decimal_from_per_thing(per_thing), Decimal::new(10, 2));
+/// ```
+fn decimal_from_per_thing<T: PerThing>(per_thing: T) -> Result<Decimal, ArithmeticError> {
+	let numerator =
+		Decimal::from_u128(per_thing.deconstruct().into()).ok_or(ArithmeticError::Overflow)?;
+	let denominator =
+		Decimal::from_u128(T::one().deconstruct().into()).ok_or(ArithmeticError::Overflow)?;
+
+	numerator.safe_div(&denominator)
+}
+
+/// Paramaters:
+/// * `lp_token_details` - Vec of tuples containing `(token_deposit, token_balance, token_weight)`
+/// * `f` - Fee
+///
+/// https://github.com/ComposableFi/composable/blob/main/rfcs/0008-pablo-lbp-cpp-restructure.md#42-liquidity-provider-token-lpt-math-updates
+/// Equation 6
+fn compute_first_deposit_lp_<T: PerThing>(
+	lp_token_details: Vec<(u128, u128, T)>,
+	_f: T,
+) -> Result<(u128, u128), ConstantProductAmmError> {
+	let k: u128 = lp_token_details.len().try_into().map_err(|_| ArithmeticError::Overflow)?;
+	let product = lp_token_details
+		.iter()
+		.try_fold(1, |product, (_d_i, b_i, _w_i)| product.safe_mul(b_i))?;
+
+	// REVIEW: How is fee applied here?
+	Ok((k.safe_mul(&product)?, 0))
+}
```

Comment:
> How is fee applied here?

How are fees calculated for first deposits? Should it be per asset? instead of a single `u128`?